packages/opencensus-instrumentation-http2/src/http2.ts

```diff
@@ -0,0 +1,283 @@
+      if (propagation) {
+        propagation.inject(setter, span.spanContext);
+      }
+
+      request.on('response', (responseHeaders: http2.IncomingHttpHeaders) => {
+        span.addAttribute(
+            Http2Plugin.ATTRIBUTE_HTTP_STATUS_CODE,
+            `${responseHeaders[':status']}`);
+        span.status = plugin.traceStatus(+responseHeaders[':status']);
+      });
+
+      request.on('end', () => {
+        const userAgent =
+            (headers['user-agent'] || headers['User-Agent'] || null) as string;
+
+        span.addAttribute(
+            Http2Plugin.ATTRIBUTE_HTTP_HOST, plugin.clientUrl.host);
+        span.addAttribute(
+            Http2Plugin.ATTRIBUTE_HTTP_METHOD, `${headers[':method']}`);
+        span.addAttribute(
+            Http2Plugin.ATTRIBUTE_HTTP_PATH, `${headers[':path']}`);
+        span.addAttribute(
+            Http2Plugin.ATTRIBUTE_HTTP_ROUTE, `${headers[':path']}`);
+        span.addAttribute(Http2Plugin.ATTRIBUTE_HTTP_USER_AGENT, userAgent);
+
+        span.addMessageEvent(
+            'MessageEventTypeSent', uuid.v4().split('-').join(''));
+
+        span.end();
+      });
+
+      request.on('error', (err: Error) => {
+        span.addAttribute(HttpPlugin.ATTRIBUTE_HTTP_ERROR_NAME, err.name);
+        span.addAttribute(HttpPlugin.ATTRIBUTE_HTTP_ERROR_MESSAGE, err.message);
+
+        span.end();
+      });
+
+      return request;
+    };
+  }
+
+  patchCreateServer() {
+    const plugin = this;
+    return (original: CreateServerFunction): types.Func<http2.Http2Server> => {
+      return function patchedCreateServer(): http2.Http2Server {
+        plugin.server = original.apply(this, arguments);
+        shimmer.wrap(
+            plugin.server.constructor.prototype, 'emit', plugin.patchEmit());
+
+        shimmer.unwrap(plugin.moduleExports, 'createServer');
+        shimmer.unwrap(plugin.moduleExports, 'createSecureServer');
+        plugin.patchedCreateServer = true;
+
+        return plugin.server;
+      };
+    };
+  }
+
+  patchEmit() {
+    const plugin = this;
+    return (original: RequestFunction): types.Func<http2.ClientHttp2Stream> => {
+      return function patchedEmit(
+                 event: string, stream: http2.ServerHttp2Stream,
+                 headers: http2.IncomingHttpHeaders): http2.ClientHttp2Stream {
+        if (event !== 'stream') {
+          return original.apply(this, arguments);
+        }
+
+        const propagation = plugin.tracer.propagation;
+        const getter = {
+          getHeader(name: string) {
+            return headers[name];
+          }
+        } as types.HeaderGetter;
+
+        const traceOptions = {
+          name: headers[':path'],
+          type: 'SERVER',
+          spanContext: propagation ? propagation.extract(getter) : null
+        } as types.TraceOptions;
+
+        let statusCode: number = null;
+        const originalRespond = stream.respond;
+        stream.respond = function(this: http2.Http2Stream) {
```

Comment:
Can you add a comment here as to why you are wrapping `respond`, and why it only needs to be wrapped once? Where does it get called? Is it possible for it to be called more than once?

Reply:
Added a comment explaining. Please let me know if it doesn't make it clearer.